rsg: improve argument generation

madelynnblue opened the pull request:
More random ints, bytes, and strings. This now catches the REPEAT bug,
so it must be ignored.

Fixes #15191

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/15672)
<!-- Reviewable:end -->

justinj commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/internal/rsg/rsg.go, line 170 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15672#-KjF3R9iejOrBu1RB-e4:-KjF3R9iejOrBu1RB-e5:bxn6ycs) ([raw file](https://github.com/cockroachdb/cockroach/blob/e26d2a1f107e50996e62ef0ac0ff96408ef8855a/pkg/internal/rsg/rsg.go#L170)):*
> ```Go
> 	case 7:
> 		i = r.src.Int63()
> 		if r.src.Intn(2) == 1 {
> ```

I don't know how common this kind of pattern is elsewhere in RSG, but maybe something like this would simplify some of this logic:

```
i *= r.Sample(1, -1)
```
where `Sample` is something like,
```
func (r *RSG) Sample(args... int) int {
	return args[r.src.Intn(len(args))]				
}
```
but it might just add more noise if this kind of thing isn't too common.

---

*[pkg/internal/rsg/rsg.go, line 174 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15672#-KjF2sgxeB1SKVYiWHnQ:-KjF2sgxeB1SKVYiWHnR:b-5sr1ra) ([raw file](https://github.com/cockroachdb/cockroach/blob/e26d2a1f107e50996e62ef0ac0ff96408ef8855a/pkg/internal/rsg/rsg.go#L174)):*
> ```Go
> 		}
> 	case 8:
> 		for v := r.src.Intn(10) + 1; v > 0; v-- {
> ```

just to confirm, this is used instead of `r.src.Intn(1000000000)` in order to get a more exponential distribution of numbers?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15672#-:-KjF6OX_XSXDKHyj0ptO:bnfp4nl)*
<!-- Sent from Reviewable.io -->

mjibson commented:
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/internal/rsg/rsg.go, line 170 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15672#-KjF3R9iejOrBu1RB-e4:-KjF92Bci-EEqlA69l8a:b-9ac6wg) ([raw file](https://github.com/cockroachdb/cockroach/blob/e26d2a1f107e50996e62ef0ac0ff96408ef8855a/pkg/internal/rsg/rsg.go#L170)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

I don't know how common this kind of pattern is elsewhere in RSG, but maybe something like this would simplify some of this logic:

```
i *= r.Sample(1, -1)
```
where `Sample` is something like,
```
func (r *RSG) Sample(args... int) int {
	return args[r.src.Intn(len(args))]				
}
```
but it might just add more noise if this kind of thing isn't too common.


</blockquote></details>

Yeah, I don't think we need that quite yet.

---

*[pkg/internal/rsg/rsg.go, line 174 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15672#-KjF2sgxeB1SKVYiWHnQ:-KjF94PVKc9Ohnyi830d:b-t0axno) ([raw file](https://github.com/cockroachdb/cockroach/blob/e26d2a1f107e50996e62ef0ac0ff96408ef8855a/pkg/internal/rsg/rsg.go#L174)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

just to confirm, this is used instead of `r.src.Intn(1000000000)` in order to get a more exponential distribution of numbers?
</blockquote></details>

Correct. We want an even distribution from all the orders of magnitude.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15672)*
<!-- Sent from Reviewable.io -->

madelynnblue committed:
rsg: improve argument generation

More random ints, bytes, and strings. This now catches the REPEAT bug,
so it must be ignored.

Fixes #15191
